src/db/getCategoryMessages.js

```diff
@@ -1,26 +1,58 @@
+const _ = require('highland')
+const { tap } = require('tinyfunk')
+
 const debug = require('../lib/debug').extend('db')
 const parseMessage = require('../lib/parseMessage')
 
 const sql = 'SELECT * FROM message_store.get_category_messages($1, $2, $3, $4, $5, $6, $7)'
 
 // getCategoryMessages :: Object -> Stream Message
-const getCategoryMessages = ({ queryS }) => opts => {
+const getCategoryMessages = ({ query }) => opts => {
   debug('getting category messages: %o', opts)
 
-  const vals = [
-    opts.category,
-    opts.position,
-    opts.batchSize,
-    opts.correlation,
-    opts.groupMember,
-    opts.groupSize,
-    opts.condition,
-  ]
-
-  const stream = queryS(sql, vals)
-  const result = stream.map(parseMessage)
-  result.close = stream.close
-  return result
+  let {
+    batchSize,
+    category,
+    condition,
+    correlation,
+    groupMember,
+    groupSize,
+    position
+  } = opts
+
+  const paginate = (push, next) => {
+    const vals = [
+      category,
+      position,
+      batchSize,
+      correlation,
+      groupMember,
+      groupSize,
+      condition
+    ]
+
+    const send = rows => {
+      if (rows.length) {
+        debug('category messages found: %o', { category, count: rows.length })
+        push(null, rows)
+        next()
+      } else {
+        debug('end of category: %o', { category })
+        push(null, _.nil)
+      }
+    }
+
+    query(sql, vals).then(send, push)
```

Comment:
`.catch(push)`?

Reply:
Is that not equivalent to supplying `push` as the second argument to `.then()`?

Reply:
It sure is. I forget that `then` accepts two args. 😞

Reply:
Actually... it's not.  I tested in a repl.  If you `first().then(second, fail)`, then `fail` is not executed if `second`, er... fails.  So I've swapped it for `.catch()` instead.